secio/src/lib.rs

```diff
@@ -100,3 +108,132 @@ impl Digest {
+        self.sign_ecdsa(message)
+    }
+
+    /// Constructs a signature for `msg` using the secret key `sk`
+    fn sign_ecdsa<T: AsRef<[u8]>>(&self, message: T) -> Result<Vec<u8>, Self::Error>;
+
+    /// Creates a new public key from the [`KeyProvider`].
+    fn pubkey(&self) -> Self::Pubkey;
+}
+
+/// Public key for KeyProvider
+pub trait Pubkey: Send + Sync + 'static {
+    /// Error
+    type Error: Into<crate::error::SecioError>;
+    /// Checks that `sig` is a valid ECDSA signature for `msg` using the public
+    /// key `pubkey`.
+    fn verify_ecdsa<T: AsRef<[u8]>, F: AsRef<[u8]>>(&self, message: T, signature: F) -> bool;
+
+    /// serialized key into a bytes
+    fn serialize(&self) -> Vec<u8>;
+
+    /// Recover public key from slice
+    fn from_slice<T: AsRef<[u8]>>(key: T) -> Result<Self, Self::Error>
+    where
+        Self: Sized;
+}
+
+impl KeyProvider for SecioKeyPair {
+    type Error = error::SecioError;
+    type Pubkey = secp256k1_compat::PublicKey;
+
+    fn sign_ecdsa<T: AsRef<[u8]>>(&self, message: T) -> Result<Vec<u8>, Self::Error> {
+        let msg = match crate::secp256k1_compat::message_from_slice(message.as_ref()) {
+            Ok(m) => m,
+            Err(_) => {
+                log::debug!("message has wrong format");
+                return Err(error::SecioError::InvalidMessage);
+            }
+        };
+        let signature = match self.inner {
+            KeyPairInner::Secp256k1 { ref private } => crate::secp256k1_compat::sign(&msg, private),
+        };
+
+        Ok(crate::secp256k1_compat::signature_to_vec(signature))
+    }
+
+    fn pubkey(&self) -> Self::Pubkey {
+        match self.inner {
+            KeyPairInner::Secp256k1 { ref private } => {
+                crate::secp256k1_compat::from_secret_key(private)
+            }
+        }
+    }
+}
+
+impl Pubkey for secp256k1_compat::PublicKey {
+    type Error = error::SecioError;
+    fn verify_ecdsa<T: AsRef<[u8]>, F: AsRef<[u8]>>(&self, message: T, signature: F) -> bool {
+        let signature = crate::secp256k1_compat::signature_from_der(signature.as_ref());
+        let msg = crate::secp256k1_compat::message_from_slice(message.as_ref());
+
+        if let (Ok(signature), Ok(message)) = (signature, msg) {
+            if !crate::secp256k1_compat::verify(&message, &signature, self) {
+                log::debug!("failed to verify the remote's signature");
+                return false;
+            }
+        } else {
+            log::debug!("remote's secp256k1 signature has wrong format");
+            return false;
+        }
+        true
+    }
+
+    fn serialize(&self) -> Vec<u8> {
+        crate::secp256k1_compat::serialize_pubkey(self)
+    }
+
+    fn from_slice<T: AsRef<[u8]>>(key: T) -> Result<Self, Self::Error> {
+        crate::secp256k1_compat::pubkey_from_slice(key.as_ref())
+            .map_err(|_| crate::error::SecioError::SecretGenerationFailed)
+    }
+}
+
+impl KeyProvider for () {
```

Comment:
I think this implementation is only for unit test, suggest to add `#[cfg(test)]` or implement a  NoopKeyProvider struct.